controllers/controllers/workloads/cfpackage_controller_test.go

```diff
@@ -57,4 +60,59 @@ var _ = Describe("CFPackageReconciler Integration Tests", func() {
 			}))
 		})
 	})
+
+	When("a CFPackage is deleted", func() {
+		BeforeEach(func() {
+			cfPackage = BuildCFPackageCRObject(cfPackageGUID, cfSpace.Status.GUID, cfAppGUID)
+		})
+
+		JustBeforeEach(func() {
+			Expect(k8sClient.Create(context.Background(), cfPackage)).To(Succeed())
+
+			// wait for package to have reconciled at least once
+			Eventually(func(g Gomega) {
+				g.Expect(k8sClient.Get(context.Background(), client.ObjectKeyFromObject(cfPackage), cfPackage)).To(Succeed())
+				g.Expect(cfPackage.Finalizers).ToNot(BeEmpty())
+			}).Should(Succeed())
+
+			Expect(k8sClient.Delete(context.Background(), cfPackage)).To(Succeed())
+		})
+
+		It("deletes itself and the corresponding source image", func() {
+			Eventually(func(g Gomega) {
+				g.Expect(imageDeleter.DeleteCallCount()).To(Equal(1))
+			}).Should(Succeed())
+
+			_, ref := imageDeleter.DeleteArgsForCall(0)
+			Expect(ref).To(Equal("PACKAGE_IMAGE"))
+
+			Eventually(func(g Gomega) {
+				g.Expect(k8sClient.Get(context.Background(), client.ObjectKeyFromObject(cfPackage), cfPackage)).To(MatchError(ContainSubstring("not found")))
+			}).Should(Succeed())
+		})
+
+		When("the package doesn't have an image set", func() {
+			BeforeEach(func() {
+				cfPackage.Spec.Source.Registry.Image = ""
+			})
+
+			It("doesn't try to delete any image", func() {
+				Consistently(func(g Gomega) {
+					g.Expect(imageDeleter.DeleteCallCount()).To(Equal(0))
+				}).Should(Succeed())
+			})
+		})
+
+		When("deletion fails", func() {
+			BeforeEach(func() {
+				imageDeleter.DeleteReturns(errors.New("oops"))
+			})
+
+			It("ignores the errors and finishes finalization", func() {
```

Comment:
Cool I think this is good. I was worried about an external resource outage blocking finalization. Any ideas around how to make this scenario easier for operators to detect though? It reminds me of the "orphaned blob" problem that CF for VMs found itself dealing with.